common/src/Renderer/BrushRenderer.cpp

```diff
@@ -460,65 +457,82 @@ namespace TrenchBroom {
             auto& facesSortedByTex = brushCache.cachedFacesSortedByTexture();
             const size_t facesSortedByTexSize = facesSortedByTex.size();
 
-            std::shared_ptr<TextureToBrushIndicesMap> faceVboPtr = \
-                (renderType == Filter::RenderOpacity::Opaque) ? m_opaqueFaces : m_transparentFaces;
-
             size_t nextI;
             for (size_t i = 0; i < facesSortedByTexSize; i = nextI) {
                 const Assets::Texture* texture = facesSortedByTex[i].texture;
 
-                size_t indexCount = 0;
+                size_t opaqueIndexCount = 0;
+                size_t transparentIndexCount = 0;
 
                 // find the i value for the next texture
-                for (nextI = i + 1; nextI < facesSortedByTexSize && facesSortedByTex[nextI].texture == texture; ++nextI) {
-                }
+                for (nextI = i + 1; nextI < facesSortedByTexSize && facesSortedByTex[nextI].texture == texture; ++nextI) {}
 
                 // process all faces with this texture (they'll be consecutive)
                 for (size_t j = i; j < nextI; ++j) {
                     const BrushRendererBrushCache::CachedFace& cache = facesSortedByTex[j];
                     if (cache.face->isMarked()) {
                         assert(cache.texture == texture);
-                        indexCount += triIndicesCountForPolygon(cache.vertexCount);
+                        if (m_transparent || cache.face->hasAttribute(Model::TagAttributes::Transparency)) {
+                            transparentIndexCount += triIndicesCountForPolygon(cache.vertexCount);
+                        } else {
+                            opaqueIndexCount += triIndicesCountForPolygon(cache.vertexCount);
+                        }
                     }
                 }
 
-                // there may be no marked faces with this texture
-                if (indexCount == 0) {
-                    continue;
-                }
+                if (transparentIndexCount > 0) {
```

Comment:
We could maybe put this bit that's copied and pasted twice into a lambda, and execute it twice, once for the opaque faces and once for the translucent. OTOH, I'm OK with leaving it as it is now and refactoring it next time I work on this code..

Reply:
Yeah, I considered that, but it seemed to have so many dependencies. But a lambda might work fine! I'll do it on the other PR.